extends：继承

异常处理

throws：表示该方法可能会抛出的异常，允许throws后面跟着多个异常类型
throw：用于抛出异常。当方法在执行过程中遇到异常情况时，将异常信息封装为异常对象，然后throw

```Java
public   class  ThrowAndThrows  {
    /**
     * 计算一个数字的平方根
     * @param nStr    以字符串的形式提供数字
     * @return    返回平方根
     * @throws Exception    当用户输入的字符串为空，
     * 或者字符串无法转换成数字，或者转换成的数字小于0，都会抛出异常
     */
    public static double sqrt(String nStr) throws Exception {
        if (nStr == null) {
            //用trow关键字抛出异常，当异常被抛出时，程序会跳出该方法
            throw new Exception("输入的字符不能为空！");
        }
        double n = 0;
        try {
            n = Double.parseDouble(nStr);
        } catch(NumberFormatException e) {
            //将parseDouble方法可能抛出的异常NumberFormatException捕获，
            //然后将捕获的异常重新封装并输出
            throw new Exception("输入的字符串必须能够转化成数字！", e);
        }
        if (n < 0 ){
            throw new Exception("输入的字符串转化成的数字必须大于0！");
        }
        return Math.sqrt(n);
    }
    
    public static void main(String[] args) throws Exception {
        try{
            ThrowAndThrows.sqrt("-124.56");
        } catch(Exception e) {
            //将sqrt方法声明的可能抛出的Exception异常捕获
            //打印捕获的异常的堆栈信息，从堆栈信息中可以发现异常发生的位置和原因
            System.out.println("Got a Exception：" + e.getMessage());
            e.printStackTrace();
            throw e;    //不做进一步处理，将异常向外抛出
        }
        
        //将sqrt声明了可能会抛出的异常向外抛出，必须在方法声明中使用throws
        ThrowAndThrows.sqrt("-124.56");
    }

}
————————————————
版权声明：本文为CSDN博主「zhouyong80」的原创文章，遵循CC 4.0 BY-SA版权协议，转载请附上原文出处链接及本声明。
原文链接：https://blog.csdn.net/zhouyong80/article/details/1907799
```

